snapshot/windows/windows.go

```diff
@@ -40,45 +79,304 @@ func NewSnapshotter(root string) (snapshot.Snapshotter, error) {
+	ctx, t, err := o.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return err
+	}
+	defer t.Rollback()
+	return storage.WalkInfo(ctx, fn)
+}
+
+func (o *snapshotter) mounts(s storage.Snapshot) []mount.Mount {
+	var (
+		roFlag string
+	)
+
+	if s.Kind == snapshot.KindView {
+		roFlag = "ro"
+	} else {
+		roFlag = "rw"
+	}
+
+	var mounts []mount.Mount
+
+	mounts = append(mounts, mount.Mount{
+		Source: o.getSnapshotDir(s.ID),
+		Type:   "windows-layer",
+		Options: []string{
+			roFlag,
+		},
+	})
+
+	for _, ID := range s.ParentIDs {
+		mounts = append(mounts, mount.Mount{
+			Source: o.getSnapshotDir(ID),
+			Type:   "windows-layer",
+		})
+	}
+
+	return mounts
+}
+
+func (o *snapshotter) getSnapshotDir(id string) string {
+	return filepath.Join(o.root, "snapshots", id)
+}
+
+func (o *snapshotter) createSnapshot(ctx context.Context, kind snapshot.Kind, key, parent string, opts []snapshot.Opt) ([]mount.Mount, error) {
+	ctx, t, err := o.ms.TransactionContext(ctx, true)
+	if err != nil {
+		return nil, err
+	}
+
+	s, err := storage.CreateSnapshot(ctx, kind, key, parent, opts...)
+	if err != nil {
+		if rerr := t.Rollback(); rerr != nil {
+			log.G(ctx).WithError(rerr).Warn("Failure rolling back transaction")
+		}
+		return nil, errors.Wrap(err, "failed to create snapshot")
+	}
+
+	switch kind {
+	case snapshot.KindView:
+		var parentID string
+		if len(s.ParentIDs) != 0 {
+			parentID = s.ParentIDs[0]
+		}
+		if err := hcsshim.CreateLayer(o.info, s.ID, parentID); err != nil {
+			if rerr := t.Rollback(); rerr != nil {
+				log.G(ctx).WithError(rerr).Warn("Failure rolling back transaction")
+			}
+			return nil, errors.Wrap(err, "failed to create layer")
+		}
+	case snapshot.KindActive:
+		layerChain := o.parentIDsToLayerChain(s.ParentIDs)
+
+		var parentPath string
+		if len(layerChain) != 0 {
+			parentPath = layerChain[0]
+		}
+
+		if err := hcsshim.CreateSandboxLayer(o.info, s.ID, parentPath, layerChain); err != nil {
+			if rerr := t.Rollback(); rerr != nil {
+				log.G(ctx).WithError(rerr).Warn("Failure rolling back transaction")
+			}
+			return nil, errors.Wrap(err, "failed to create sandbox layer")
+		}
+
+		// TODO(darrenstahlmsft): Allow changing sandbox size
+		//	if err := hcsshim.ExpandSandboxSize(o.info, id, size); err != nil {
+		//		return err
+		//	}
+	}
+
+	if err := t.Commit(); err != nil {
+		return nil, errors.Wrap(err, "commit failed")
+	}
+
+	return o.mounts(s), nil
+}
+
+func (o *snapshotter) parentIDsToLayerChain(parentIDs []string) []string {
+	var layerChain []string
+	for _, ID := range parentIDs {
+		layerChain = append(layerChain, o.getSnapshotDir(ID))
+	}
+	return layerChain
+}
+
+// getFileSystemType obtains the type of a file system through GetVolumeInformation
+// https://msdn.microsoft.com/en-us/library/windows/desktop/aa364993(v=vs.85).aspx
+func getFileSystemType(drive string) (fsType string, hr error) {
```

Comment:
Do we want to upstream this to continuity?

Reply:
That sounds good to me.